NASA-PDS/validate
# Fix bug for label with FITS file with multiple Header tags before and…

… after Table_Binary tag

1. Add test resources : src/test/resources/github416
2. New class to remove comments from XML file  : src/main/java/gov/nasa/pds/tools/util/DocumentUtil.java
3. New class to count number of header tags from file before Table_ tags : src/main/java/gov/nasa/pds/tools/util/TableUtil.java
4. Fix bug logic when summing up header size by being aware of how many headers before the Table_ tag : src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java
5. Add new type FIELDS_MISMATCH to count errors in regression test  : src/test/java/cucumber/StepDefs.java
6. Add 4 new regression tests : src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/416 validate 2.1.0 indicates the table offset is not correct when validating a binary table inside a FITS

Closes https://github.com/nasa-pds/validate/issues/416 validate 2.1.0 indicates the table offset is not correct when validating a binary table inside a FITS

A bug occured due the label has many "Header" tags but only two were meant for the "Table_Binary".  The third "Header" was meant for the "Array_3D_Image".
The fix was to count the number of "Header" tagbs before the "Table_" field.  With this fix, the header size is now calculated correctly.

Tested in DEV:

Since user provided two labels that were valid but validate module marked them as invalid, there are 4 tests to cover 2 valid cases and 2 invalidate cases.

Case 1:

Running validate against first valid label.  No error expected.
```
% validate -R pds4.label -r report_github416_label_valid_1.json  -s json -t src/test/resources/github416/mix_raw_calib_mixs-c_sw_offset_table_20160301.xml
```

Case 2:
Running validate against first label with the offset still inside the header.  Expect one error for error.table.fields_mismatch type.

```
% validate -R pds4.label -r report_github416_label_invalid_1.json  -s json -t src/test/resources/github416/mix_raw_calib_mixs-c_sw_offset_table_20160301_invalid.xml
```

Case 3:
Running validate against a second valid label.  No error expected.

```
% validate -R pds4.label -r report_github416_label_valid_2.json  -s json -t src/test/resources/github416/phe_misc_temperature_reference_20190524.xml
```

Case 4:
Running validate against a second label with the offset still inside the header.  Expect one error for error.table.fields_mismatch type.

```
% validate -R pds4.label -r report_github416_label_invalid_2.json  -s json -t src/test/resources/github416/phe_misc_temperature_reference_20190524_invalid.xml
```
